man/runsvdir.8
.\" Automatically generated by Pandoc 3.1.11.1
.\"
.TH "runsvdir" "8" 
.SH NAME
runsvdir \- starts and monitors a collection of runsv(8) processes
.SH SYNOPSIS
\f[B]runsvdir\f[R] [\-P] \f[I]dir\f[R] [ \f[I]log\f[R] ]
.SH DESCRIPTION
\f[I]dir\f[R] must be a directory.
\f[I]log\f[R] is a space holder for a readproctitle log, and must be at
least seven characters long or absent.
.PP
\f[B]runsvdir\f[R] starts a \f[B]runsv\f[R](8) process for each
subdirectory, or symlink to a directory, in the services directory
\f[I]dir\f[R], up to a limit of 1000 subdirectories, and restarts a
\f[B]runsv\f[R](8) process if it terminates.
\f[B]runsvdir\f[R] skips subdirectory names starting with dots.
\f[B]runsv\f[R](8) must be in \f[B]runsvdir\f[R]\[aq]s PATH.
.PP
At least every five seconds \f[B]runsvdir\f[R] checks whether the time
of last modification, the inode, or the device, of the services
directory \f[I]dir\f[R] has changed.
If so, it re\-scans the services directory, and if it sees a new
subdirectory, or new symlink to a directory, in \f[I]dir\f[R], it starts
a new \f[B]runsv\f[R](8) process; if \f[B]runsvdir\f[R] sees a
subdirectory being removed that was previously there, it sends the
corresponding \f[B]runsv\f[R](8) process a TERM signal, stops monitoring
this process, and so does not restart the \f[B]runsv\f[R](8) process if
it exits.
.PP
If the \f[I]log\f[R] argument is given to \f[B]runsvdir\f[R], all output
to standard error is redirected to this \f[I]log\f[R], which is similar
to the daemontools\[aq] \f[B]readproctitle\f[R] log.
To see the most recent error messages, use a process\-listing tool such
as \f[B]ps\f[R](1).
\f[B]runsvdir\f[R] writes a dot to the readproctitle log every 15
minutes so that old error messages expire.
.SH OPTIONS
.TP
\f[B]\-P\f[R]
use \f[B]setsid\f[R](2) to run each \f[B]runsv\f[R](8) process in a new
session and separate process group.
.SH SIGNALS
If \f[B]runsvdir\f[R] receives a TERM signal, it exits with 0
immediately.
.PP
If \f[B]runsvdir\f[R] receives a HUP signal, it sends a TERM signal to
each \f[B]runsv\f[R](8) process it is monitoring and then exits with
111.
.SH SEE ALSO
sv(8), runsv(8), runsvchdir(8), runit(8), runit\-init(8), chpst(8),
svlogd(8), setsid(2)
.PP
https://smarden.org/runit/
.SH AUTHOR
Gerrit Pape <pape\[at]smarden.org>


package/CHANGES
  * chpst.c, man/chpst.8: add new option -N: with -012 reopen to /dev/null;
    -2: dup() and coe() stderr, restore on failed exec (github/pilcrow).
  * man/runsvdir.8: typo.
  * doc/*.html: add color-scheme.
  * runsvdir.c: use inotify(7) (if available) to watch services directory;
    remove overcautious delay; cleanup removed services more quickly.
  * Makefile, TARGETS: add test for inotify_init() system call.
  * hasinotify.h1, hasinotify.h2, tryinotify.c: new.
  * runsvdir.c: add selfpipe; buffer and optimize readproctitle log; close
    unused filedescriptor.
  * chpst.c: -F, -I: move signal handler functions to another ifdef; do not
    relay SIGCHLD also in main pid.
  * sv.c: status: if log/supervise/ can't be accessed, don't print confusing
    additional line after warning message (introduced with 5fe1bc7).
  * etc/debian: rename to etc/sysv.
  * etc/freebsd: rename to etc/bsd.
  * etc/openbsd, etc/freebsd/getty-ttyv4, etc/debian/getty-tty5: remove;
    obsolete.
  * doc/replaceinit.html: revise sections "Replacing init on *BSD" and
    "Replacing sysvinit on GNU/Linux".
  * doc/install.html: how to install as non-root user; optionally use cc on
    *BSD.
  * man/chpst.8: document new -F and -I options.
  * chpst.c: -F option implies -I (thx github/andy-bower).
  * chpst.c: add new option -F: fghack using unshare(); chpst.c: add new
    option -I: run another init using unshare().
  * Makefile, TARGETS: add test for unshare() system call (_GNU_SOURCE).
  * hasunshare.h1, hasunshare.h2, tryunshare.c: new.
  * conf-cc: trim.
  * doc/install.html, doc/replaceinit.html, doc/runlevels.html,
    doc/useinit.html: strip '^# ' from code blocks for multiline copy&paste.
  * man/sv.8, doc/dependencies.html: rename check script to program.
  * doc/runscripts.html: clarify outdated status of the collection.
  * doc/*.html, man/*.8: replace; converted from markdown source.
  * svlogd.c, man/svlogd.8: -LL: print warning on data loss.
  * svlogd.c, man/svlogd.8: add option -L, lossy (thx github/teslamotors).
  * package/upgrade: remove utmpset program on upgrade.
  * doc/replaceinit.html: remove getty finish scripts that use utmpset;
    etc/debian/getty-tty5/finish, etc/freebsd/getty-ttyv4/finish,
    etc/openbsd/getty-ttyC4/finish: remove, obsolete.
  * doc/index.html, man/runit-init.8, man/runit.8, man/runsv.8,
    man/runsvdir.8: remove reference to utmpset program.
  * utmpset.c, utmpset.check, utmpset.dist, man/utmpset.8, tryuwtmp.c,
    tryuwtmpx.c, uw_tmp.h?: remove, obsolete.
  * Makefile, package/commands, src/Makefile, TARGETS: remove utmpset and
    utmpx sysdep.

runit 2.3.1
Sat, 07 Feb 2026 11:11:01 +0100
  * doc/dependencies.html: wording; cruft.
  * runsv.c, man/runsv.8: run control/? with service process id as argument.
  * sv.c: always write to control pipe, do not look at the status.
  * svlogd.c: fix uninitialized pointer read with -v and log directory
    mode r--.
  * Makefile, TARGETS, byte.h: add byte_zero to byte.a.
  * byte_zero.c: new.
  * runsv.c: omit stat, pid in ./supervise/ if RUNSV_STATUS_BINARYONLY is
    defined.
  * chpst.dist, sv.dist, svlogd.dist: adapt.
  * buffer_put.c: fix compilation warning: -Wempty-body (github/clan).
  * buffer_put.c, byte.h, byte_copy.c, byte_diff.c: fix compilation
    warnings: -Wdiscarded-qualifiers (github/clan).
  * sv.c: fix compilation warning: -Wdangling-else (github/clan).
  * byte_chr.c, byte_copy.c, byte_cr.c, byte_diff.c, byte_rchr.c, chpst.c,
    str_chr.c, str_len.c, sv.c, svlogd.c: fix compilation warnings:
    -Wmisleading-indentation (github/clan).
  * TARGETS: remove prot.o (github/clan).
  * doc/benefits.html: spelling.
  * doc/faq.html: cleanup, spelling.
  * doc/usedietlibc.html: remove.
  * doc/index.html: remove references to dietlibc, spelling.
  * man/runsv.8, doc/runsv.8.html: sync; clarify wording wrt log service and
    custom control char o; style.
  * man/runit.8, man/sv.8, man/svlogd.8: typo, groff.

runit 2.3.0
Sun, 09 Nov 2025 11:07:17 +0100
  * TARGETS: remove hasshsgr.h.
  * runsvctrl.*, runsvstat.*, svwaitdown.*, svwaitup.*: remove, obsolete;
    Makefile, TARGETS: remove runsvctrl, runsvstat, svwaitup, svwaitdown.
  * runsv.c: fix typo in pathname (thx github/plorenzo19).
  * sv.c, man/sv.8: fix "sv check" to behave as documented (thx Buck Evan,
    James Byrne).
  * hasflock.h1: add #include <unistd.h>.
  * doc/dependencies.html: rewrite; new initial terse version.
  * doc/index.html: replace broken link to sysvinit.
  * doc/runscripts.html: remove broken links and reference to mailing list.
  * doc/replaceinit.html: switch http links to https.
  * svlogd.c: never prune the in-progress .u (thx Mike Pomraning for the
    draft patch).
  * sv.c: fix regression introduced with 5fe1bc7: exit code of sv status in
    "lsb" mode is not 3 in all cases when the service is "down" (thx Fabian
    Ruff, David Mountney for the patch).
  * runsvchdir.c, man/runsvchdir.8: the environment variable $RUNSVDIR
    now overrides the default working directory /etc/runit/runsvdir/
    (github/plorenzo19).
  * runit.c: really restart stage 2 if it crashed.
  * alloc.c, alloc_re.c, buffer_0.c, byte.h, byte_chr.c, byte_copy.c,
    byte_cr.c, byte_diff.c, byte_rchr.c, select.h2, wait.h, wait_nohang.c,
    wait_pid.c: C23: swap ANSI for remaining K&R-style declarations
    (github/andy-bower).
  * alloc.h, buffer.c, buffer.h, buffer_get.c, buffer_put.c, buffer_write.c,
    byte.h, pathexec_env.c, runit.c, runsv.c, runsvdir.c, select.h1,
    select.h2, sig.c, sig.h, sig_catch.c, svlogd.c, svlogd.dist, wait.h:
    fix build with gcc15: in C23 f() means f(void), consequently signal
    handler functions now are f(int), alloc_free() parameter needs cast,
    buffer_unixwrite() removes "const" from parameter.
  * doc/index.html: repair or remove broken links in list of distributions
    using runit, add Debian GNU/Hurd.
  * doc/*, man/*: switch all cr.yp.to http links to https.

src/runsvdir.c
#include "hasinotify.h"
#ifdef HASINOTIFY
#include <sys/inotify.h>
#include <limits.h>
#endif
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <signal.h>
#include "direntry.h"
#include "strerr.h"
#include "error.h"
#include "sgetopt.h"
#include "wait.h"
#include "env.h"
#include "open.h"
#include "pathexec.h"
#include "fd.h"
#include "byte.h"
#include "str.h"
#include "coe.h"
#include "iopause.h"
#include "sig.h"
#include "ndelay.h"

#define USAGE " [-P] [-L fifo] dir [log]"
#define VERSION "$Id$"

#define MAXSERVICES 1000

char *progname;
char *svdir;
unsigned long dev =0;
unsigned long ino =0;
struct {
  unsigned long dev;
  unsigned long ino;
  int pid;
  int isgone;
} sv[MAXSERVICES];
int svnum =0;
int check =1;
int selfpipe[2];
char *rplog =0;
int rploglen;
int logpipe[2] ={-1};
char *fifolog =0;
int fifo[2] ={-1};
#ifdef HASINOTIFY
#define INBUFSIZE (sizeof(struct inotify_event) +NAME_MAX +1 > 256 ? \
                   sizeof(struct inotify_event) +NAME_MAX +1 : 256)
#define IONUM 2
int watch[3];
#else
#define INBUFSIZE 256
#define IONUM 1
#endif
char inbuf[INBUFSIZE];
iopause_fd io[IONUM +1];
struct taia stamplog;
int exitsoon =0;
int pgrp =0;

void usage () { strerr_die4x(1, "usage: ", progname, USAGE, "\n"); }
void fatal(char *m1, char *m2) {
  strerr_die6sys(100, "runsvdir ", svdir, ": fatal: ", m1, m2, ": ");
}
void warn(char *m1, char *m2) {
  strerr_warn6("runsvdir ", svdir, ": warning: ", m1, m2, ": ", &strerr_sys);
}
void warn3x(char *m1, char *m2, char *m3) {
  strerr_warn6("runsvdir ", svdir, ": warning: ", m1, m2, m3, 0);
}
void s_term(int unused) { exitsoon =1; write(selfpipe[1], "", 1); }
void s_hangup(int unused) { exitsoon =2; write(selfpipe[1], "", 1); }
void s_child(int unused) { write(selfpipe[1], "", 1); }

void runsv(int no, char *name) {
  int pid;

  if ((pid =fork()) == -1) {
    warn("unable to fork for ", name);
    return;
  }
  if (pid == 0) {
    /* child */
    char *prog[3];

    prog[0] ="runsv";
    prog[1] =name;
    prog[2] =0;
    sig_uncatch(sig_pipe);
    sig_uncatch(sig_hangup);
    sig_unblock(sig_hangup);
    sig_uncatch(sig_term);
    sig_unblock(sig_term);
    sig_uncatch(sig_child);
    sig_unblock(sig_child);
    if (pgrp) setsid();
    pathexec_run(*prog, prog, (char* const*)environ);
    fatal("unable to start runsv ", name);
  }
  sv[no].pid =pid;
}

void runsvdir() {
  DIR *dir;
  direntry *d;
  int i;
  struct stat s;

  if (! (dir =opendir("."))) {
    warn("unable to open directory ", svdir);
    return;
  }
  for (i =0; i < svnum; i++) sv[i].isgone =1;
  errno =0;
  while ((d =readdir(dir))) {
    if (d->d_name[0] == '.') continue;
    if (stat(d->d_name, &s) == -1) {
      warn("unable to stat ", d->d_name);
      errno =0;
      continue;
    }
    if (! S_ISDIR(s.st_mode)) continue;
    for (i =0; i < svnum; i++) {
      if ((sv[i].ino == s.st_ino) && (sv[i].dev == s.st_dev)) {
        sv[i].isgone =0;
        if (! sv[i].pid) runsv(i, d->d_name);
        break;
      }
    }
    if (i == svnum) {
      /* new service */
      if (svnum >= MAXSERVICES) {
        warn3x("unable to start runsv ", d->d_name, ": too many services.");
        continue;
      }
      sv[i].ino =s.st_ino;
      sv[i].dev =s.st_dev;
      sv[i].pid =0;
      sv[i].isgone =0;
      svnum++;
      runsv(i, d->d_name);
      check =1;
    }
  }
  if (errno) {
    warn("unable to read directory ", svdir);
    closedir(dir);
    check =1;
    return;
  }
  closedir(dir);

  /* SIGTERM removed runsv's */
  for (i =0; i < svnum; i++) {
    if (! sv[i].isgone) continue;
    if (sv[i].pid) kill(sv[i].pid, SIGTERM);
    sv[i--] =sv[--svnum];
    check =1;
  }
}

int setup_fifo() {
  if ((fifo[0] =open_read(fifolog)) == -1) {
    warn("unable to open fifo: ", fifolog);
    return(0);
  }
  if ((fifo[1] =open_write(fifolog)) == -1) {
    warn("unable to open fifo for writing: ", fifolog);
    close(fifo[0]);
    fifo[0] =-1;
    return(0);
  }
  coe(fifo[0]);
  coe(fifo[1]);
  ndelay_on(fifo[0]);
  ndelay_on(fifo[1]);
  return(1);
}

int setup_log() {
  if (pipe(logpipe) == -1) {
    warn3x("unable to create pipe for log.", 0, 0);
    return(-1);
  }
  coe(logpipe[1]);
  coe(logpipe[0]);
  ndelay_on(logpipe[0]);
  ndelay_on(logpipe[1]);
  if (fd_move(2, logpipe[1]) == -1) {
    warn3x("unable to set filedescriptor for log.", 0, 0);
    return(-1);
  }
  io[IONUM].fd =logpipe[0];
  io[IONUM].events =IOPAUSE_READ;
  if (rplog) taia_now(&stamplog);
  return(1);
}
#ifdef HASINOTIFY
unsigned int watch_inotify() {
  int w;

  if ((w =inotify_add_watch(io[1].fd, svdir, IN_DONT_FOLLOW|
          IN_DELETE_SELF|IN_MOVE_SELF)) == -1)
    return(0);
  if (watch[0] != w) inotify_rm_watch(io[1].fd, watch[0]);
  watch[0] =w;
  if ((w =inotify_add_watch(io[1].fd, svdir,
          IN_DELETE_SELF|IN_MOVE_SELF|IN_CREATE|IN_DELETE|IN_MOVE)) == -1)
    return(0);
  if (watch[1] != w) inotify_rm_watch(io[1].fd, watch[1]);
  watch[1] =w;
  if (watch[0] != watch[1]) {
    if ((w =readlink(svdir, inbuf, 256)) != -1) {
      if (w < 256) {
        inbuf[w] =0;
        if (*inbuf == '/') {
          if ((w =inotify_add_watch(io[1].fd, inbuf, IN_DONT_FOLLOW|
                  IN_MASK_ADD|IN_DELETE_SELF|IN_MOVE_SELF)) == -1)
            return(0);
          if (watch[2] == w) return(1);
          if (watch[2] != -1) inotify_rm_watch(io[1].fd, watch[2]);
          watch[2] =(watch[1] == w) ? -1 : w;
          return(1);
        }
      }
      else
        warn3x("unable to readlink ", svdir, ": name too long");
    }
    else
      if (errno != EINVAL) warn("unable to readlink ", svdir);
  }
  if (watch[2] != -1) inotify_rm_watch(io[1].fd, watch[2]);
  watch[2] =-1;
  return(1);
}
#endif

int main(int argc, char **argv) {
  struct stat s;
  time_t mtime =0;
  int wstat;
  int curdir;
  int pid;
  struct taia deadline;
  struct taia now;
  struct taia stampcheck;
  char ch;
  int i;

  progname =*argv;
  while ((i =getopt(argc, (char* const*)argv, "PL:")) != opteof) {
    switch(i) {
    case 'L': fifolog =(char*)optarg; break;
    case 'P': pgrp =1; break;
    case '?': usage();
    }
  }
  argv +=optind;
  if (! argv || ! *argv) usage();
  svdir =*argv++;

  if (pipe(selfpipe) == -1) fatal("unable to create selfpipe", 0);
  coe(selfpipe[0]);
  coe(selfpipe[1]);
  ndelay_on(selfpipe[0]);
  ndelay_on(selfpipe[1]);
  io[0].fd =selfpipe[0];
  io[0].events =IOPAUSE_READ;

  if (argv && *argv) {
    rplog =*argv;
    if ((rploglen =str_len(rplog)) < 7) {
      warn3x("log argument too short: ", "readproctitle log disabled.", 0);
      rplog =0;
    }
    rplog +=5; rploglen -=5;
  }
  if (rplog || fifolog)
    if (setup_log() != 1) {
      warn3x("log service disabled.", 0, 0);
      rplog =fifolog =0;
    }
  if (fifolog) { sig_ignore(sig_pipe); setup_fifo(); }
  if ((curdir =open_read(".")) == -1)
    fatal("unable to open current directory", 0);
  coe(curdir);
#ifdef HASINOTIFY
  if ((io[1].fd =inotify_init()) == -1)
    fatal("unable to initialize inotify instance", 0);
  coe(io[1].fd);
  ndelay_on(io[1].fd);
  io[1].events =IOPAUSE_READ;
  if ((watch[0] =inotify_add_watch(io[1].fd, svdir, IN_DONT_FOLLOW|
          IN_DELETE_SELF|IN_MOVE_SELF)) == -1)
    fatal("unable to add watch to inotify instance", 0);
  if ((watch[1] =inotify_add_watch(io[1].fd, svdir,
          IN_DELETE_SELF|IN_MOVE_SELF|IN_CREATE|IN_DELETE|IN_MOVE)) == -1)
    fatal("unable to add watch to inotify instance", 0);
  watch[2] =-1;
#endif
  sig_block(sig_term);
  sig_catch(sig_term, s_term);
  sig_block(sig_hangup);
  sig_catch(sig_hangup, s_hangup);
  sig_block(sig_child);
  sig_catch(sig_child, s_child);
  taia_now(&stampcheck);

  for (;;) {
    /* collect children */
    for (;;) {
      if ((pid =wait_nohang(&wstat)) <= 0) break;
      for (i =0; i < svnum; i++) {
        if (pid == sv[i].pid) {
          /* runsv has gone */
          sv[i].pid =0;
          check =1;
          break;
        }
      }
    }

    taia_now(&now);
    if (now.sec.x < (stampcheck.sec.x -3)) {
      /* time warp */
      warn3x("time warp: resetting time stamp.", 0, 0);
      taia_now(&stampcheck);
      taia_now(&now);
      if (rplog) taia_now(&stamplog);
    }
    if (taia_less(&now, &stampcheck) == 0) {
      /* wait at least a second */
      taia_uint(&deadline, 1);
      taia_add(&stampcheck, &now, &deadline);
      
      if (stat(svdir, &s) != -1) {
        if (check || \
            s.st_mtime != mtime || s.st_ino != ino || s.st_dev != dev) {
          /* svdir modified */
#ifdef HASINOTIFY
          if (!watch_inotify())
            warn("unable to add watch to inotify instance", 0);
#endif
          if (chdir(svdir) != -1) {
            mtime =s.st_mtime;
            dev =s.st_dev;
            ino =s.st_ino;
            check =0;
            runsvdir();
            while (fchdir(curdir) == -1) {
              warn("unable to change directory, pausing", 0);
              sleep(5);
            }
          }
          else
            warn("unable to change directory to ", svdir);
        }
      }
      else
        warn("unable to stat ", svdir);
    }

    if (rplog)
      if (taia_less(&now, &stamplog) == 0) {
        for (i =1; i < rploglen; i++) rplog[i -1] =rplog[i];
        rplog[rploglen -1] ='.';
        taia_uint(&deadline, 900);
        taia_add(&stamplog, &now, &deadline);
      }
    taia_uint(&deadline, check ? 1 : 5);
    taia_add(&deadline, &now, &deadline);
    if (rplog && taia_less(&stamplog, &deadline)) deadline =stamplog;

    sig_unblock(sig_hangup);
    sig_unblock(sig_term);
    sig_unblock(sig_child);
    iopause(io, IONUM + (logpipe[0] > -1 ? 1 : 0), &deadline, &now);
    sig_block(sig_child);
    sig_block(sig_term);
    sig_block(sig_hangup);

    if (io[0].revents) while (read(selfpipe[0], &ch, 1) == 1) {}
#ifdef HASINOTIFY
    if (io[1].revents) {
      check =1;
      while (read(io[1].fd, inbuf, sizeof(inbuf)) > 0) {}
    }
#endif
    if ((logpipe[0] > -1) && io[IONUM].revents) {
      if (fifolog && (fifo[0] == -1))
        if (setup_fifo())
          write(fifo[1], "runsvdir: warning: logs are incomplete\n", 39);
      while ((i =read(logpipe[0], inbuf, 256)) > 0) {
        if (rplog) {
          int j;
          if (i < rploglen)
            for (j =0; j < rploglen -i; ++j) rplog[j] =rplog[j +i];
          j =(i > rploglen) ? rploglen : i;
          byte_copy(rplog +rploglen -j, j, inbuf +i -j);
        }
        if (fifo[0] > -1)
          if ((write(fifo[1], inbuf, i) == -1) && (errno == EAGAIN)) {
            close(fifo[1]); close(fifo[0]); fifo[0] =-1;
          }
      }
    }

    switch(exitsoon) {
    case 1:
      _exit(0);
    case 2:
      for (i =0; i < svnum; i++) if (sv[i].pid) kill(sv[i].pid, SIGTERM);
      _exit(111);
    }
  }
  /* not reached */
  _exit(0);
}
